Index: src/umu_commander/__init__.py
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.BaseRevisionTextPatchEP
<+>VERSION = \"v1.5.6\"\n
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/umu_commander/__init__.py b/src/umu_commander/__init__.py
--- a/src/umu_commander/__init__.py	(revision ebc968e623b6ee38499de12ff2a04c4bf2311db2)
+++ b/src/umu_commander/__init__.py	(date 1754484120387)
@@ -1,1 +1,0 @@
-VERSION = "v1.5.6"
